update-changelog/src/helpers.ts

```diff
@@ -0,0 +1,81 @@
+import * as semver from "semver";
+import moment from "moment";
+import * as fs from "fs";
+
+const changelogRegex = /^## (?<currentVersion>[^\n]*)[^#]*(?<sections>.*?)\n## (?<prevVersion>[^ ]*)/gms;
+const sectionsRegex = /### (?<sectionName>[^\n]*)(?<sectionContent>[^#]*)/gm;
+
+export function processChangelog(changelog: string): { updatedChangelog: string; newVersion: string } {
+    changelogRegex.lastIndex = 0;
+    sectionsRegex.lastIndex = 0;
+
+    const changelogMatch = changelogRegex.exec(changelog);
+    if (!changelogMatch || !changelogMatch.groups) {
+        throw new Error(`Failed to match changelog: ${changelog}`);
+    }
+
+    const prevVersion = changelogMatch.groups["prevVersion"];
+    let newVersion = prevVersion;
+
+    const parsedPrevVersion = semver.parse(prevVersion);
+    if (!parsedPrevVersion) {
+        throw new Error(`Failed to parse previous version: ${prevVersion}`);
+    }
+    if (parsedPrevVersion.prerelease.length) {
+        newVersion = semver.inc(prevVersion, "prerelease")!;
+    } else {
+        let sectionMatch: RegExpExecArray | null;
+        while ((sectionMatch = sectionsRegex.exec(changelogMatch.groups["sections"]))) {
```

Comment:
If I understand correctly you don't need a loop here as the first element of the returned array is always the matched text. So that should suffice, unless I misinterpreted the doc of `RegExp.prototype.exec()`.

Reply:
This is the example I used: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#finding_successive_matches

Reply:
Misinterpretation from my side. I understand what they meant with
>The returned array has the matched text as the first item, and then one item for each parenthetical capture group of the matched text.